keps/sig-api-machinery/3920-finish-feature-flags/README.md

```diff
@@ -0,0 +1,951 @@
+* It's mostly unsafe to turn a feature on, use it, and turn it back off. This is
+  because state stored while the feature was on is not cleaned up. So instead of
+  going back into a pristine state, the cluster goes into a third state,
+  on-with-dangling-references. This third state is tested less well than either
+  the on or off states.
+* Because of this, it's risky to try alpha features with important clusters.
+* It's not very useful or convenient to make disposable clusters to try alpha
+  features.
+* Because of that, it's extremely hard to get robust feedback about features
+  while they are in alpha and easy to change.
+* There is no place in the code for devs to put turn on / turn off code, even if
+  they wanted to.
+* Likewise there is no place to put upgrade or downgrade code. Many efforts have
+  used built-in resources instead of CRDs simply because there is no place to
+  put CRD install code.
+* It is difficult to know if a feature has been turned on for every relevant
+  binary. Yes, there are metrics, but you have to know which binaries are
+  relevant for what features and your particular cluster's metrics collection
+  system.
+* It is especially difficult to know if use of a feature at some point makes
+  cluster upgrade unsafe now; for example, an alpha object being written while
+  the feature was briefly on which is incompatable with the default-on beta.
+* It is not easy to test the feature on -> use -> feature off lifecycle.
+
+### Goals
+
+Primary goals:
+* Minimize codebase churn, existing feature authors do little-to-no work
+* Expose the feature states in use in the cluster
+* Make it possible to get the cluster back into the "off" state
+* Expose the "off but dangling references / uses" state
+* Make it easy to add enablement and disablement code
+* Make it easy to add upgrade and downgrade code
+* Make it easy to tell if it is not safe to upgrade / downgrade due to feature
+  states
+* A test ensures that feature on -> use -> feature off works for all features
+  (i.e. it's easy for feature authors to be included in this test)
+
+Secondary goals:
+* Third parties should be able to use the API and libraries we build for their
+  own features.
+* Set feature states centrally (on apiserver command line) -- may not apply to
+  all features
+* Permit the platform admin to permit selected features to be dynamically
+  settable, by api clients
+
+Follow-up goals:
+* When a feature is about a new field, remove it from OpenAPI specs when the
+  feature is off.
+* For features that possibly could support it, add a canarying mechanism so that
+  the feature enablement doesn't break the whole cluster at once when there is a
+  problem.
+
+### Non-Goals
+
+* The feature system will not permit passing configuration, since that's already
+  done in a variety of ways today. Therefore only features that don't take
+  config can be dynamic. Maybe this can be addressed in a follow up.
+
+## Proposal
+
+### High-level design
```

Comment:
It would be a lot easier to understand if you discussed the roles that will be interacting with the Feature API - platform admin, cluster admin, feature developer, feature consumer (user), others? - and then explained the uses for those different roles.

As it reads now, I feel like I am missing context, at least on the first read.

Reply:
I was trying to do that and introduce them one at a time, but I can uh give a list or something?

Reply:
Yeah, something to help people build a mental model or frame it in context. Even if you define them one at a time, that could be OK, but you need to explain who they are when you introduce them, not leave it as implicit. I just felt dropped in the middle of a thought process as it is now.

Reply:
I pushed some changes, is this better?